### PR TITLE
Deprecate Trainer.training_type_plugin in favor of trainer.strategy

### LINKED ISSUE #10416: [Main Issue] Accelerator and Plugin refactor
### Motivation
Accelerator is not stable API yet, we can improve the Accelerator related logic and move towards stable Accelerator version for 1.6

### Pitch
**Steps**
1. Collective refactor #7534
   - [x] deprecate accelerator collective directly call from TTP #9677
   - [ ] collective refactor #9414
2. Move Precision Plugin into TTP #7324 
3. Move Accelerator into Strategy #10648 
4. Simplify the Spawning logic #10059 
5. [RFC] Simplifying the Accelerator Connector logic and flags (can be done in parallel with aboves) #11449  11449
6. [RFC] Revisit the inheritance of TTP #11863
 
 
 

**More details in**: [Accelerator Refactor Proposal](https://docs.google.com/document/d/1E5t8auWf5DrNHzutvMmrJC_KqBqVyZuYX0thra69Ad8/edit#heading=h.nehmb825f40s)
[updating] 

### FAQ
1. Will this be a lot of breaking changes?
Not much user facing API changes from 1,2,3,4.(Unless we found out other existing bugs during refactor) The only breaking change will be for custom plugins
5 and 6 is still RFC stage, may have breaking changes which impact user facing APIs

2. How does this impact lightningLite?
Should be helpful for lightningLite too, there maybe function refactor/simplification could happen for lightningLite. (@awaelchli  any suggestion about this part?)

### Follow up TODOs:
- [ ] check if trainer.FITTING before setting up optimizers https://github.com/PyTorchLightning/pytorch-lightning/pull/10596#discussion_r753587321

- [ ] Send a PR to  https://github.com/ray-project/ray_lightning  to update their plugins

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @tchaton @borda @kaushikb11 @ananthsub

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?
follow up for  #11120 will rebase after it get merged

No logic change, only update reference name

part of the clean up work for #10416

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]
LGTM.

Some commands that you might find useful for follow-ups

```python
$ grep -iIrn "ttp" pytorch_lightning tests | grep -vi "http"
```

```python
$ grep -iIrn "training_type" pytorch_lightning tests
```

**Comment by four4fish** [comment]
> ```python
> training_type
> ```

@carmocca Good call! Updated ttp and training_type words. The training_type pass in import will be updated when we change the file name

File: CHANGELOG.md [hunk 1]
```diff
@@ -129,12 +129,22 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 - Marked the `ResultCollection`, `ResultMetric`, and `ResultMetricCollection` classes as protected ([#11130](https://github.com/PyTorchLightning/pytorch-lightning/pull/11130))
 
 
 - DeepSpeed does not require lightning module zero 3 partitioning ([#10655](https://github.com/PyTorchLightning/pytorch-lightning/pull/10655))
 
 
+- Rename `training_type_plugin` property to `strategy` in `Trainer` and updated the references ([#11141](https://github.com/PyTorchLightning/pytorch-lightning/pull/11141))
+
+
+- Renamed `_call_ttp_hook` method to `_call_strategy_hook` in `Trainer` ([#11141](https://github.com/PyTorchLightning/pytorch-lightning/pull/11141))
```
Comment (awaelchli) [inline CHANGELOG.md:144]: Let's separate this please

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1674,41 +1672,41 @@ def _on_exception(self) -> None:
...
     @property
     def accelerator(self) -> Accelerator:
-        return self.training_type_plugin.accelerator
+        return self.strategy.accelerator
 
     @property
-    def training_type_plugin(self) -> TrainingTypePlugin:
-        return self._accelerator_connector.training_type_plugin
+    def strategy(self) -> Strategy:
+        return self._accelerator_connector.strategy
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1679]: Here we need to add the deprecation path, reroute the old to the new + test in tests/deprecated_api

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1679]: Oh totally forgot about this is trainer API, will update

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -283,58 +283,58 @@ def _handle_accelerator_and_strategy(self) -> None:
...
             )
 
-    def _set_training_type_plugin(self) -> None:
-        if isinstance(self.strategy, str) and self.strategy in TrainingTypePluginsRegistry:
-            self._training_type_plugin = TrainingTypePluginsRegistry.get(self.strategy)
-        if isinstance(self.strategy, str):
-            self.set_distributed_mode(self.strategy)
-        elif isinstance(self.strategy, Strategy):
-            self._training_type_plugin = self.strategy
+    def _set_strategy(self) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:303]: Careful, all these changes here are not directly related to the deprecation of the Trainer.ttp property. 
In the future, please remain super focused, especially when these rename-PRs become large very quickly.

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1669,41 +1667,49 @@ def _on_exception(self) -> None:
...
+
+    @property
+    def strategy(self) -> Strategy:
+        return self._accelerator_connector.strategy
 
     @property
     def training_type_plugin(self) -> Strategy:
-        return self._accelerator_connector.training_type_plugin
+        rank_zero_deprecation(
+            "`Trainer.training_type_plugin` is deprecated in v1.6 and will be removed in v2.0. Use"
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1679]: 2.0? That sounds extreme 😅

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
Comment (four4fish) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:303]: @awaelchli If we only deprecate but not update training.training_type_plugin reference to trainer.strategy everywhere, All tests will fail (like everywhere calls a trainer.fit()), or we will have to add pytest.raiseError()...
Any recommendations how to separate the deprecation and update reference?

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1679]: 1.8 then :)?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:303]: it is clear that his cannot be separated. my comment was about the highlighted part where other methods and local variables get renamed that are unrelated to the property.

Comment (four4fish) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:303]: Ohh! I see, I saw variable names in accelerator_connector all get updated, will remove these changes. I thought I separated them, thank you for reminder.

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1679]: Yes that would be the first choice imo. 
It was not agreed on what 2.0 means formally, but my understanding was that it will concern api changes that are deeply rooted in how Lightning is used today. This cannot be said about the ttp IMO, which is rarely accessed by the user. The usual 2 version deprecation process should be more than enough :)

File: docs/source/extensions/plugins.rst [hunk 5]
```diff
@@ -70,7 +70,7 @@ Expert users may choose to extend an existing plugin by overriding its methods .
                 device_ids=...,
             )
 
-or by subclassing the base classes :class:`~pytorch_lightning.plugins.training_type.Strategy` or
+or by subclassing the base classes :class:`~pytorch_lightning.plugins.training_type.training_type_plugin` or
```
Comment (akihironitta) [inline docs/source/extensions/plugins.rst:73]: I suppose this should be kept as is.
```suggestion
or by subclassing the base classes :class:`~pytorch_lightning.plugins.training_type.Strategy` or
```

**Review by tchaton — APPROVED** [review]
LGTM ! Nice progress.

File: docs/source/extensions/plugins.rst (hunk 5 above)
Comment (awaelchli) [reply docs/source/extensions/plugins.rst:73]: not yet. this was accidentally changed in another PR. The files have not been renamed yet.

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
Comment (four4fish) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: @awaelchli  All non-trainer.training_type_plugin name change has removed

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: That's great, will merge it after all classes have been renamed
